catalog/imaging/imaging-actionprovider-chip/src/main/java/org/codice/alliance/imaging/chip/actionprovider/ImagingChipActionProvider.java

```diff
@@ -32,110 +30,115 @@
-    if (resultUrl != null) {
-      return resultUrl;
-    }
-
-    resultUrl = getResultUrl(metacard, Metacard.DERIVED_RESOURCE_URI);
-    if (resultUrl != null) {
-      return resultUrl;
-    }
-
-    return SystemBaseUrl.getBaseUrl();
+    final Metacard metacard = (Metacard) input;
+
+    return Stream.<Supplier<Optional<URL>>>of(
+            () ->
+                attemptToConstructChippingUrlFromDerivedResourceAttribute(
+                    metacard, Metacard.DERIVED_RESOURCE_DOWNLOAD_URL),
+            () ->
+                attemptToConstructChippingUrlFromDerivedResourceAttribute(
+                    metacard, Metacard.DERIVED_RESOURCE_URI),
+            () -> attemptToConstructDefaultChippingUrl(metacard))
+        .map(Supplier::get)
+        .filter(Optional::isPresent)
+        .map(Optional::get)
+        .map(
+            url ->
+                Collections.singletonList(
+                    (Action) new ActionImpl(getId(), TITLE, DESCRIPTION, url)))
+        .findFirst()
+        .orElse(new ArrayList<>());
   }
 
-  @Nullable
-  private String getResultUrl(Metacard metacard, String attributeName) {
-    if (hasValueForAttribute(metacard, attributeName)) {
-      try {
-        return getBaseUrlFromAttribute(metacard, attributeName);
-      } catch (MalformedURLException e) {
-        LOGGER.debug("Could not transform derived resource download URL from string to URL", e);
+  /**
+   * If the derived resource attribute value matches the usual Alliance download URL format ("[http
+   * or https]://[derived resource host]:[derived resource port]/services/catalog/sources/[derived
+   * resource source id]/[derived resource metacard id]?transform=resource&qualifier=[original or
+   * overview]"), assume that there is a chipping URL that can be constructed from the scheme, host,
+   * port, source, and id of the value.
+   */
+  private Optional<URL> attemptToConstructChippingUrlFromDerivedResourceAttribute(
+      final Metacard metacard, final String derivedResourceAttributeName) {
+    final Optional<URI> derivedResourceUriOptional =
+        getDerivedResourceUriByQualifier(metacard, derivedResourceAttributeName, "original");
+
+    if (derivedResourceUriOptional.isPresent()) {
+      final URI derivedResourceUri = derivedResourceUriOptional.get();
+
+      final String scheme =
+          derivedResourceUri.getScheme(); // {@code null} if the scheme is undefined
+      final String host = derivedResourceUri.getHost(); // {@code null} if the host is undefined
+      final int port = derivedResourceUri.getPort(); // {@code -1} if the port is undefined
+      final String path = derivedResourceUri.getPath(); // {@code null} if the path is undefined
+      final String query = derivedResourceUri.getQuery(); // {@code null} if the query is undefined
+      if ((StringUtils.equals(scheme, "http") || StringUtils.equals(scheme, "https"))
+          && !StringUtils.isEmpty(host)
+          && port != -1
```

Comment:
Might want to allow for a case where the port is not specified.

Reply:
`getPort()` will return -1 if the port is undefined: https://docs.oracle.com/javase/7/docs/api/java/net/URL.html#getPort()